File: src/main/java/algo/backtrack/Permutations_67.java
package algo.backtrack;

import java.util.ArrayList;
import java.util.List;

public class Permutations_67 {
  public static List<List<Integer>> permute(int[] nums) {
    var result = new ArrayList<List<Integer>>();
    backtrack(nums, result, new ArrayList<>(), new boolean[nums.length]);
    return result;
  }

  private static void backtrack(int[] nums, List<List<Integer>> result, List<Integer> perm, boolean[] visited) {
    if (perm.size() == nums.length) {
      result.add(new ArrayList<>(perm));
    }
    for (var i = 0; i < nums.length; i++) {
      if (!visited[i]) {
        visited[i] = true;
        perm.add(nums[i]);
        backtrack(nums, result, perm, visited);
        visited[i] = false;
        perm.remove(perm.size() - 1);
      }
    }
  }

  public static void main(String[] args) {
//    permute(new int[]{1,2,3}).forEach(System.out::println);
    permute(new int[]{1,2,3,4,5,6,7,8,9}).forEach(System.out::println);
  }
}

File: src/main/kotlin/algo/backtrack/Permutations_67.kt
package algo.backtrack

class KtPermutations_67 {
    fun permute(nums: IntArray): List<List<Int>> {
        val all: MutableList<List<Int>> = mutableListOf()
        fun backtrack(nums: IntArray, list: MutableList<Int>, visited: BooleanArray) {
            if (list.size == nums.size) {
                all.add(list.toList())
                return
            }
            for (i in nums.indices) {
                if (visited[i]) {
                    continue
                }
                visited[i] = true
                list.add(nums[i])
                backtrack(nums, list, visited)
                visited[i] = false
                list.removeLast()
            }
        }
        backtrack(nums, mutableListOf(), BooleanArray(nums.size))
        return all
    }
}

fun main() {
    val permute = KtPermutations_67()
    println(permute.permute(intArrayOf(1,2,3)))
    println(permute.permute(intArrayOf(0,1)))
    println(permute.permute(intArrayOf(1)))
}